go/pkg/cas/upload.go

```diff
@@ -589,6 +604,144 @@ func (u *uploader) uploadBatch(ctx context.Context, reqs []*repb.BatchUpdateBlob
 	})
 }
 
+// stream uploads the item using ByteStream service.
+//
+// If the blob is already uploaded, then the function returns quickly and
+// without an error.
+func (u *uploader) stream(ctx context.Context, item *uploadItem, updateCacheStats bool) error {
+	if err := u.semByteStreamWrite.Acquire(ctx, 1); err != nil {
+		return err
+	}
+	defer u.semByteStreamWrite.Release(1)
+
+	// Allocate resources.
+	enc := zstdEncoders.Get().(*zstd.Encoder)
+	fileBuf := u.fileIOBufs.Get().([]byte)
+	streamBuf := u.streamBufs.Get().([]byte)
+	defer func() {
+		enc.Close()
+		zstdEncoders.Put(enc)
+		u.fileIOBufs.Put(fileBuf)
+		u.streamBufs.Put(streamBuf)
+	}()
+
+	// Open the item.
+	r, err := item.Open()
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+
+	// TODO(nodir): implement per-RPC timeouts. No nice way to do it.
+	rewind := false
+	return u.withRetries(ctx, func(ctx context.Context) error {
+		// Do not rewind if this is the first attempt.
+		if rewind {
+			if _, err := r.Seek(0, io.SeekStart); err != nil {
```

Comment:
I think these retries have to start at the `WriteResponse.CommitedSize` -- or you'll be writing from zero on top of partially written data on the server. See https://github.com/googleapis/googleapis/blob/master/google/bytestream/bytestream.proto#L140

Reply:
Added a TODO.